File: Leetcode/0095-unique-binary-search-trees-ii/solution.java
import java.util.*;

class TreeNode {
    int val; TreeNode left, right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left = left; this.right = right; }
}

class Solution {
    public List<TreeNode> generateTrees(int n) {
        if (n == 0) return new ArrayList<>();
        return gen(1, n);
    }
    private List<TreeNode> gen(int lo, int hi) {
        List<TreeNode> result = new ArrayList<>();
        if (lo > hi) { result.add(null); return result; }
        for (int root = lo; root <= hi; root++) {
            for (TreeNode l : gen(lo, root - 1)) {
                for (TreeNode r : gen(root + 1, hi)) {
                    result.add(new TreeNode(root, l, r));
                }
            }
        }
        return result;
    }

    static int passed = 0, failed = 0;
    static int catalan(int n) {
        if (n <= 1) return 1;
        int[] c = new int[n + 1]; c[0] = c[1] = 1;
        for (int i = 2; i <= n; i++) for (int j = 0; j < i; j++) c[i] += c[j] * c[i - 1 - j];
        return c[n];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        for (int n = 0; n <= 6; n++) {
            int got = sol.generateTrees(n).size();
            int want = (n == 0) ? 0 : catalan(n);
            if (got == want) { System.out.println("PASS: n=" + n + " → " + got); passed++; }
            else { System.out.println("FAIL: n=" + n + " got=" + got + " want=" + want); failed++; }
        }
        System.out.printf("%nResults: %d passed, %d failed, %d total%n", passed, failed, passed + failed);
    }
}

File: Leetcode/0095-unique-binary-search-trees-ii/solution.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        if n == 0: return []
        def gen(lo, hi):
            if lo > hi: return [None]
            result = []
            for root in range(lo, hi + 1):
                for L in gen(lo, root - 1):
                    for R in gen(root + 1, hi):
                        result.append(TreeNode(root, L, R))
            return result
        return gen(1, n)


def catalan(n):
    if n <= 1: return 1
    c = [1, 1]
    for i in range(2, n + 1):
        c.append(sum(c[j] * c[i - 1 - j] for j in range(i)))
    return c[n]


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    for n in range(0, 7):
        got = len(sol.generateTrees(n))
        want = 0 if n == 0 else catalan(n)
        if got == want: print(f"PASS: n={n} → {got}"); passed += 1
        else: print(f"FAIL: n={n} got={got} want={want}"); failed += 1
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
